tensorboard/loader.py

```diff
@@ -732,85 +740,367 @@ def get_next_event(self):
+    get_next_event, then no i/o should be necessary.
+
+    Returns:
+      The step count of the first summary event, or None if there
+      didn't appear to be any summary events, or the user isn't logging
+      step counts.
+
+    :rtype: int
+    """
+    if self._first_step is not None:
+      return self._first_step
+    if self._update_first_step_invocations >= EventLogReader.FIRST_STEP_PEEKS:
+      return None
+    with RecordReader(self.path) as reader:
+      for _ in range(EventLogReader.FIRST_STEP_PEEKS):
+        record = reader.get_next_record()
+        if record is None:
+          break
+        event = tf.Event()
+        event.ParseFromString(record.record)
+        self._update_first_step(event)
+    return self._first_step
+
+  def close(self):
+    """Closes event log reader if open.
+
+    Further i/o is not permitted after this method is called.
+    """
+    if self._reader is not None:
+      self._reader.close()
+      self._reader = None
+
+  def _update_first_step(self, event):
+    self._update_first_step_invocations += 1
+    # proto3 doesn't let us distinguish between absent and 0.
+    if event.step:
+      step = 0 if self._observed_zero_step else event.step
+      if self._first_step is None or step < self._first_step:
+        self._first_step = step
+    elif event.summary:
+      self._observed_zero_step = True
+
+  def __hash__(self):
+    return hash(self._key)
+
+  def __eq__(self, other):
+    return self._key == other._key
+
+  def __lt__(self, other):
+    return self._key < other._key
+
+  def __str__(self):
+    offset = self.get_offset()
+    if offset:
+      return u'EventLogReader{path=%s, offset=%d}' % (self.path, offset)
+    else:
+      return u'EventLogReader{%s}' % self.path
+
+
+@util.closeable
+@functools.total_ordering
+@six.python_2_unicode_compatible
+class RunReader(object):
+  """Utility for loading event logs into the DB.
+
+  This class merges the chain of event log files into one meaningful
+  stream of events, ordered by step or timestamp.
+
+  Fields:
+    rowid: The primary key of the corresponding row in Runs.
+    name: Display name of this run.
+  """
+
+  def __init__(self, db_conn, rowid, name):
+    """Creates new instance.
+
+    Args:
+      db_conn: A PEP 249 Connection object.
+      rowid: Primary key of run in `Runs` table, which should already
+          be inserted. This is a bit-packed int made by db.RUN_ROWID.
+      run_id: The run_id column which is encoded in the rowid bits.
+      name: Display name of run.
+
+    :type db_conn: db.Connection
+    :type rowid: int
+    :type name: str
+    """
+    self._db_conn = db_conn
+    self.rowid = db.RUN_ROWID.check(rowid)
+    self.run_id = db.RUN_ROWID.parse(rowid)[1]
+    self.name = tf.compat.as_text(name)
+    self._mark = -1
+    self._logs = []  # type: list[EventLogReader]
+    self._i = 0
+    self._entombed_progress = 0
+    self._has_new_stuff = False
```

Comment:
This is really `_has_unvisited_logs`, right? The current name suggests to me that appending to an event log would set this to `True`, but that doesn't appear to be the case.